### DISCUSSION TITLE
video 36 - routes refresh

### DISCUSSION THREAD

**Discussion by jumarko** [jacekschae/learn-pedestal-course-files#24] [Feedback]
I've been waiting for this for long  so thanks a lot!

One little thing I've noticed - you are wrapping `routes/routes` in an anonymous function.
I just used `routes/routes` as is:
```
(defn cheffy-routes [service-map]
  (assoc service-map ::http/routes
         (if (dev? service-map)
           ;; notice they used #(routes/routes) instead
           routes/routes
           (routes/routes))))
```

Is there any difference in that? Will my code be less reloadable when using this approach?

It also was not obvious me, that you can simply use a function returning routes instead of actual data structure representing the routes.
I skimmed through http://pedestal.io/guides/defining-routes and http://pedestal.io/reference/index#_routing_and_linking but couldn't didn't find any mention of that.

**Comment by jacekschae** [comment]
Hey there,
Sorry I was away for a while, with Pedestal it's the bets to look at source code: https://github.com/pedestal/pedestal/blob/d20065013abf5d3793ae5301e18a2398707fa2a9/service/src/io/pedestal/http.clj#L191-L193

Did you try to reload the code without anonymous function? Does it really work without reloading the system?

**Reply by jumarko** [reply]
Thanks for the pointer - it can be tricky to find the right place to look at (in the source code) but yeah, `default-interceptors` sounds like a good place to check.

As far as I can remember, reloading it worked well without the anonymous function.

I had other problems with reloading, notably when you change some of the interceptors.
It's not a deal breaker but it might require reloading a few different places.
Concrete and tricky example is debugging (I use cider-debugger frequently) - let's say you want to debug `transact!` here: https://github.com/jumarko/learn-pedestal/blob/main/src/main/cheffy/interceptors.clj#L30
1. You need to instrument that function (that means cider will replace the function definition with a new function that installs breakpoints)
2. You then need to reload the definition of `transact-interceptor` (https://github.com/jumarko/learn-pedestal/blob/main/src/main/cheffy/interceptors.clj#L57-L60), but not the whole file because that would replace previous instrumentation
3. Then, you need to reload [recipes routes](https://github.com/jumarko/learn-pedestal/blob/main/src/main/cheffy/recipes_with_interceptors.clj) to make sure the updated version of `transact-interceptor` is used.

So it kinda works but it's still cumbersome, especially if there's a few levels of indirection between in the routes implementation.
